scripts/install-nix-from-closure.sh

```diff
@@ -1,149 +1,635 @@
+                if ! sudo sh -c "$cmd" && print "Exiting privileged mode." ; then
+                    >&2 notice "
+
+    Password don't match or user $USER not in sudoers.
+    "
+                fi
+            else
+                print "'sudo' seems not to be installed"
             fi
-            added=1
-            break
+
         fi
-    done
 
-fi
+        # If dest still not created (with user or sudo) - use su to access root
+        if [ ! -d "$dest" ] ; then
+            print "Trying to get privileges using su and root password"
+            if ! su root sh -c "$cmd" && print "Exiting priviliged mode."; then
+                exit 1 # Look into the trap
+            fi
+        fi
+    }
+
+    createStore() {
+        trap createStoreRevert 1
+        createStoreRevert() {
+            error "
+
+    Script tried to create $dest/store , but it already exists.
+    Something is wrong with the process.
+    Please gather information of your setup conditions,
+    search does BUG exists make further actions.
+    "
+        }
+        mkdir -p "$dest"/store
+    }
+
+    copyNix() {
+        printf "copying Nix to %s..." "${dest}/store"
+
+        for i in $(cd "$self/store" >/dev/null && echo ./*); do
+            printf "." >&2
+            i_tmp="$dest/store/$i.$$"
+            if [ -e "$i_tmp" ]; then
+                rm -rf "$i_tmp"
+            fi
+            if ! [ -e "$dest/store/$i" ]; then   # If $dest/store/directory exists
+                cp -Rp "$self/store/$i" "$i_tmp" # Copy $self/store/directory to $dest/store/directory.PID
+                chmod -R a-w "$i_tmp"            # Remove write permissions in the whole $dest/store/directory.PID
+                chmod +w "$i_tmp"                # Add write permission to $dest/store/directory.PID
+                mv "$i_tmp" "$dest/store/$i"     # Place $dest/store/directory.PID as $dest/store/directory/directory.PID
+                chmod -w "$dest/store/$i"        # Remove write permissions for owner of $dest/store/directory/, but leave permissions of others untouched.
+            fi
+        done
+        printf '\n'
+    }
 
-if [ -z "$added" ]; then
-    cat >&2 <<EOF
+    initializeDB() {
+        echo "initialising Nix database..." >&2
+        if ! $nix/bin/nix-store --init; then
+            error "
 
-Installation finished!  To ensure that the necessary environment
-variables are set, please add the line
+    Failed to initialize the Nix database
+    "
+        fi
 
-  . $p
+        if ! "$nix/bin/nix-store" --load-db < "$self/.reginfo"; then
+            error "
 
-to your shell profile (e.g. ~/.profile).
-EOF
-else
-    cat >&2 <<EOF
+    Unable to register valid paths
+    "
+        fi
+    }
+
+    setupProfile() {
+        # Modifying current runtime environment variables to use Nix. Persistent changes configured later.
+        print "Launching $nix/etc/profile.d/nix.sh"
+        . "$nix/etc/profile.d/nix.sh"
```

Comment:
This line doesn't impact the user, and is internal to the script. I wouldn't log about it.

Reply:
Did you ever had an error that come up. And you can not determine directly where it come from. And you try to find a place where it was from, because no one made a message log in that source section.

Since script executes other file, it is better to state where other script borders is.

So we guide user through stdout where what was is launched.

So user directly knows what source file to read.

I care more about users have their hand on pulse what is going on in the shell script. Which they probably run as superuser.
And script to be easy enough to understand and if something - make pull requests to.

Reply:
It is for the purpose that this script informs which scrip it runs, so if that script just throws the plain uninformative error to stderr - user knows where to look and if something make a more to the point of it issue.

Oh, seems I stopped care much at this point. I calmed and decided to see if I merge it in this release, and then forget about all this.

There were several installers written at the time I informed and asked to merge any of my small fixes to Nix installer, or this cumulative PRs.

All commits cleanly show what is a rewrite of what lines. It is one-to-one comparable to the original. I spent so much time even on making ideal git history for review, four days probably only on doing the ideal git history.

I'd wanted to create the `--quiet` mode. But if this merged and then have a small PRs to the master.

While we are talking here about lines - people reported a number of issues more.

Reply:
If users not read the output - it is no biggy.
It worked fine, you skipped reading everything from the terminal.

Important info is at the end. And it is colored.

But if it is unattended, containerized or scripted install - because of detailed output there is a good log.

Yes the most frequent user would not ignore the messages. But if he has problems - he would reread wat was the autput on the screen.

That is why it is so informative.

Reply:
@wmertens you are right.

AFAIR now if it is returns error code - error code returns into the main scrip and the main script also returns error code. So it is no biggy, and before error we said that we run the `nix.sh`.

I just belive here that other scrips work, I don't try to control other scripts here, worry about other scrips, while even this code out of the tree.